perspective.el

```diff
@@ -968,6 +990,249 @@ perspective beginning with the given letter."
+transient."
+  (cond
+    ;; base case 1
+    ((not (consp entry))
+     entry)
+    ;; base case 2
+    ((atom (cdr entry))
+     entry)
+    ;; leaf: modify this
+    ((eq 'leaf (car entry))
+     (lexical-let ((leaf-props (cdr entry)))
+       (cons 'leaf
+             (cl-loop for prop in leaf-props
+                   collect (if (not (eq 'buffer (car prop)))
+                               prop
+                             (lexical-let ((bn (cadr prop)))
+                               (if (member bn valid-buffers)
+                                   prop
+                                 (cons 'buffer
+                                       (cons (lexical-let ((scratch-persp (format "*scratch* (%s)" persp)))
+                                               (if (get-buffer scratch-persp)
+                                                   scratch-persp
+                                                 "*scratch*"))
+                                             (cddr prop))))))))))
+    ;; recurse
+    (t (cons (car entry) (cl-loop for e in (cdr entry)
+                               collect (persp--state-window-state-massage e persp valid-buffers))))))
+
+(defun persp--state-frame-data ()
+  (cl-loop for frame in (frame-list)
+           if (frame-parameter frame 'persp--hash) ; XXX: filter non-perspective-enabled frames
+           collect (with-selected-frame frame
+                     (lexical-let ((persps-in-frame (make-hash-table :test 'equal))
+                                   (persp-names-in-order (persp-names)))
+                       (cl-loop for persp in persp-names-in-order do
+                                (unless (persp-killed-p (gethash persp (perspectives-hash)))
+                                  (with-perspective persp
+                                    (lexical-let* ((buffers
+                                                    (cl-loop for buffer in (persp-buffers (persp-curr))
+                                                             if (persp--state-interesting-buffer-p buffer)
+                                                             collect (buffer-name buffer)))
+                                                   (windows
+                                                    (cl-loop for entry in (window-state-get (frame-root-window) t)
+                                                             collect (persp--state-window-state-massage entry persp buffers))))
+                                      (puthash persp
+                                               (make-persp--state-single
+                                                :buffers buffers
+                                                :windows windows)
+                                               persps-in-frame)))))
+                       (make-persp--state-frame
+                        :persps persps-in-frame
+                        :order persp-names-in-order)))))
```

Comment:
sames... maybe I'm just agitating for a `TODO: refactor / clean?` comment?

Reply:
There was documented here: https://github.com/gcv/perspective-el/blob/4a0fe25/perspective.el#L1000-L1027

I'm certainly open to suggestions, but I don't see any way to simplify the code. It just takes Emacs' representation of window layouts, does a bit of transformation on unpersistable buffers, puts the result in a struct, and writes the struct to disk.